File: app/models/category.py
"""
Category & Subcategory Models
Hierarchical task categorization: Category -> Subcategory
"""

from datetime import datetime
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Boolean, Text
from sqlalchemy.orm import relationship
from app.core.database import Base


class Category(Base):
    __tablename__ = "categories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False, unique=True)
    description = Column(Text, nullable=True)
    icon = Column(String(10), nullable=True)
    color = Column(String(7), nullable=True)
    sort_order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    subcategories = relationship("Subcategory", back_populates="category", cascade="all, delete-orphan",
                                  order_by="Subcategory.sort_order")

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "icon": self.icon,
            "color": self.color,
            "sort_order": self.sort_order,
            "is_active": self.is_active,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "subcategories": [sc.to_dict() for sc in self.subcategories] if self.subcategories else []
        }


class Subcategory(Base):
    __tablename__ = "subcategories"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=True)
    category_id = Column(Integer, ForeignKey("categories.id", ondelete="CASCADE"), nullable=False)
    sort_order = Column(Integer, default=0)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    category = relationship("Category", back_populates="subcategories")

    def to_dict(self):
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "category_id": self.category_id,
            "sort_order": self.sort_order,
            "is_active": self.is_active,
            "created_at": self.created_at.isoformat() if self.created_at else None,
        }

File: mobile/lib/models/category.dart
class Category {
  final int id;
  final String name;
  final String? icon;
  final String? color;
  final List<Subcategory> subcategories;
  final int taskCount;

  Category({
    required this.id,
    required this.name,
    this.icon,
    this.color,
    this.subcategories = const [],
    this.taskCount = 0,
  });

  factory Category.fromJson(Map<String, dynamic> json) => Category(
        id: json['id'] ?? 0,
        name: json['name'] ?? '',
        icon: json['icon'],
        color: json['color'],
        subcategories: (json['subcategories'] as List?)
                ?.map((s) => Subcategory.fromJson(s))
                .toList() ??
            [],
        taskCount: json['task_count'] ?? 0,
      );
}

class Subcategory {
  final int id;
  final String name;
  final int taskCount;

  Subcategory({required this.id, required this.name, this.taskCount = 0});

  factory Subcategory.fromJson(Map<String, dynamic> json) => Subcategory(
        id: json['id'] ?? 0,
        name: json['name'] ?? '',
        taskCount: json['task_count'] ?? 0,
      );
}
